tools/torch/trace_clip.py

```diff
@@ -0,0 +1,123 @@
+# line 187 should be changed to:
+# embeddings = embeddings + self.position_embedding(self.position_ids.long())
+
+
+import torch, transformers, torchvision
+import numpy as np
+from torch import nn
+from transformers import CLIPVisionModelWithProjection, TensorType
+# import torchvision.transforms.v2 as T
+import torchvision.transforms as T
+from torchvision.transforms import InterpolationMode
+from typing import Tuple
+import argparse
+import logging
+
+class CLIPWrapper(nn.Module):
+  def __init__(self, visionmodel):
+    super(CLIPWrapper, self).__init__()
+    self.visionmodel = visionmodel
+    self.visionmodel.eval()
+    # If you prefer to script the wrapper, you can uncomment the following to internally trace the visionmodel's forward method
+    # and also remove torch.jit.script from the self.transforms
+    # self.visionmodel = torch.jit.trace(self.visionmodel.forward, example_kwarg_inputs={'pixel_values': torch.rand([1, 3, 336, 336])})
+
+    self.transforms = torch.jit.script(nn.Sequential(
+      T.ConvertImageDtype(torch.float),
+      T.Resize(size=[336,], interpolation=InterpolationMode.BICUBIC, antialias=True), # Resize to 336 on shortest edge
+      T.CenterCrop([336, 336]), # Center crop a square of 336x336
+      T.Normalize(mean=[0.48145466, 0.4578275, 0.40821073], std=[0.26862954, 0.26130258, 0.27577711])
+    ))
+
+  def forward(self, x: torch.Tensor):
+    with torch.no_grad():
+      x = self.transforms(x)
+      return self.visionmodel(pixel_values=x)[0]
+
+
+parser = argparse.ArgumentParser(description="Trace CLIP-based vision models from pytorch-transformers (only tested on openai/clip-vit-large-patch14-336 and its derivatives)")
+parser.add_argument(
+  '--model',
+  type=str,
+  help="Model to trace",
+  default="openai/clip-vit-large-patch14-336",
+)
+parser.add_argument(
+  '--cache-dir',
+  type=str,
+  help="Cache dir for HuggingFace models. Leave unset to use default",
+  default=None,
+)
+parser.add_argument(
+  '-v', '--verbose',
+  action='store_true',
+  help="Set logging level to INFO",
+)
+parser.add_argument(
+  '-o', '--output-dir',
+  type=str,
+  help="Output directory for traced models",
+  default=".",
+)
+args = parser.parse_args()
+
+if args.verbose:
+  logging.basicConfig(level=logging.INFO)
+
+logging.info(f"pytorch version {torch.__version__}, from {torch.__file__}")
+logging.info(f"transformers version {transformers.__version__}, from {transformers.__file__}")
+logging.info(f"torchvision version {torchvision.__version__}, from {torchvision.__file__}")
+
+
+# use cache dir if you want to specify a custom huggingface cache dir
+visionmodel = CLIPVisionModelWithProjection.from_pretrained(args.model, torchscript=True, cache_dir=args.cache_dir)
+
+model = CLIPWrapper(visionmodel)
+model.eval()
+
+# Uncomment if you prefer to script the wrapper
+# traced_model = torch.jit.script(model, torch.rand([1, 3, 1366, 1024]))
```

Comment:
This should be an option? only if torch.jit.script works correctly of course

Reply:
If it's just for debugging, you can tell it in the comment

Reply:
It does work, I can make it an option 👍 I wasn't sure if it was worth leaving it in since it seems to behave the same either way.